packages/hardhat-test-helpers/src/chai/matchers/bigNumber.ts

```diff
@@ -0,0 +1,264 @@
+  Assertion.overwriteMethod("below", ltFunction);
+  Assertion.overwriteMethod("lt", ltFunction);
+  Assertion.overwriteMethod("lessThan", ltFunction);
+
+  const gteFunction = override("gte", "be at least", "be below", utils);
+  Assertion.overwriteMethod("least", gteFunction);
+  Assertion.overwriteMethod("gte", gteFunction);
+  Assertion.overwriteMethod("greaterThanOrEqual", gteFunction);
+
+  const lteFunction = override("lte", "be at most", "be above", utils);
+  Assertion.overwriteMethod("most", lteFunction);
+  Assertion.overwriteMethod("lte", lteFunction);
+  Assertion.overwriteMethod("lessThanOrEqual", lteFunction);
+
+  Assertion.overwriteMethod("within", overrideWithin(utils));
+
+  Assertion.overwriteMethod("closeTo", overrideCloseTo(utils));
+}
+
+type Methods = "eq" | "gt" | "lt" | "gte" | "lte";
+
+function throwIfUnsafe(i: number) {
+  if (typeof i === "number" && !Number.isSafeInteger(i)) {
+    throw new RangeError(
+      `Cannot compare to unsafe integer ${i}. Consider using BigInt(${i}) instead. For more details, see https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number/isSafeInteger`
+    );
+  }
+}
+
+function override(
+  method: Methods,
+  name: string,
+  negativeName: string,
+  utils: Chai.ChaiUtils
+) {
+  return (_super: (...args: any[]) => any) =>
+    overwriteBigNumberFunction(method, name, negativeName, _super, utils);
+}
+
+function normalize(
+  source:
+    | number
+    | bigint
+    | typeof BNType
+    | typeof EthersBigNumberType
+    | typeof BigNumberJsType
+    | string
+): bigint {
+  if (isEthersBigNumber(source)) {
+    return BigInt(source.toString());
+  } else if (isBN(source)) {
+    return BigInt(source.toString());
+  } else if (isBigNumberJsBigNumber(source)) {
+    return BigInt(source.toString(10));
+  } else if (
+    typeof source === "string" ||
+    typeof source === "number" ||
+    typeof source === "bigint"
+  ) {
+    return BigInt(source);
+  } else {
+    throw new Error(`cannot convert ${typeof source} to BigNumber`);
+  }
+}
+
+function isBigNumber(source: any): boolean {
+  return (
+    typeof source === "bigint" ||
+    isEthersBigNumber(source) ||
+    isBN(source) ||
+    isBigNumberJsBigNumber(source)
+  );
+}
+
+function overwriteBigNumberFunction(
+  functionName: Methods,
+  readableName: string,
+  readableNegativeName: string,
+  _super: (...args: any[]) => any,
+  chaiUtils: Chai.ChaiUtils
+) {
+  return function (this: Chai.AssertionStatic, ...args: any[]) {
+    const [actualArg] = args;
+    const expectedFlag = chaiUtils.flag(this, "object");
+    if (chaiUtils.flag(this, "doLength") && isEthersBigNumber(actualArg)) {
+      // TODO: consider whether we really need this case (support for eg
+      // `expect('foo').to.have.length.of.at.least(BigInt(2));`, and if so then
+      // get it tested and generalize it to work with all the different
+      // BigNumbers.
```

Comment:
I think we do need this. It could happen because of a test like:

```js
const tokens = await contract.getTokens()
const numberOfTokens = await contract.getTokensCount()
expect(tokens).to.have.length(numberOfTokens)
```

Reply:
Addressed in https://github.com/NomicFoundation/hardhat/pull/2428/commits/bb29b72ea39b990792a3c23acc3d2f95ae6b2030